Libraries/LibMedia/Audio/PlaybackStreamWasapi.cpp

```diff
@@ -0,0 +1,502 @@
+constexpr GUID PlaybackSessionGUID = { // 22f2ca89-210a-492c-a0aa-f25b1d2f33a1
+    0x22f2ca89,
+    0x210a,
+    0x492c,
+    { 0xa0, 0xaa, 0xf2, 0x5b, 0x1d, 0x2f, 0x33, 0xa1 }
+};
+
+struct Task {
+    enum class Kind : u8 {
+        Play,
+        DrainAndSuspend,
+        DiscardAndSuspend,
+    };
+
+    void resolve(Optional<AK::Duration> time)
+    {
+        promise.visit(
+            [](Empty) { VERIFY_NOT_REACHED(); },
+            [&](NonnullRefPtr<Core::ThreadedPromise<void>>& promise) {
+                promise->resolve();
+            },
+            [&](NonnullRefPtr<Core::ThreadedPromise<AK::Duration>>& promise) {
+                promise->resolve(time.release_value());
+            });
+    }
+    void reject(Error& error)
+    {
+        promise.visit(
+            [](Empty) { VERIFY_NOT_REACHED(); },
+            [&error](auto& promise) {
+                promise->reject(move(error));
+            });
+    }
+    Variant<Empty, NonnullRefPtr<Core::ThreadedPromise<void>>, NonnullRefPtr<Core::ThreadedPromise<AK::Duration>>> promise;
+    Kind kind;
+};
+
+class ComHelper {
+public:
+    ~ComHelper()
+    {
+        if (initialized)
+            CoUninitialize();
+    }
+    bool initialized = false;
+};
+
+static thread_local ComHelper com_helper {};
+
+struct PlaybackStreamWasapi::AudioState : public AtomicRefCounted<PlaybackStreamWasapi::AudioState> {
+    AudioState();
+    ~AudioState();
+
+    ComPtr<IMMDeviceEnumerator> enumerator;
+    ComPtr<IMMDevice> device;
+    ComPtr<IAudioClient> audio_client;
+    ComPtr<IAudioRenderClient> render_client;
+    ComPtr<IAudioStreamVolume> audio_stream_volume;
+    ComPtr<IAudioClock> clock;
+
+    NonnullOwnPtr<WAVEFORMATEXTENSIBLE> wave_format;
+    UINT32 buffer_frame_count;
+    HANDLE buffer_event;
+
+    PlaybackStreamWasapi::AudioDataRequestCallback data_request_callback;
+    Function<void()> underrun_callback;
+
+    Threading::Mutex task_queue_mutex;
+    Queue<Task> task_queue;
+    // FIXME: Create a owning handle type to be shared in the codebase
+    HANDLE task_event;
+
+    bool playing = false;
+    bool drain_and_suspend = false;
+    Atomic<bool> exit_requested = false;
+
+    static int render_thread_loop(AudioState* state);
+    RefPtr<Threading::Thread> audio_thread;
+    RefPtr<Core::ThreadedPromise<AK::Duration>> resume_promise;
+    RefPtr<Core::ThreadedPromise<void>> suspend_promise;
+
+    // I doubt we will have anyone with more than 7.1
+    Vector<float, 8> channel_volumes;
+    UINT64 audio_client_frequency;
+};
+
+PlaybackStreamWasapi::AudioState::AudioState()
+    : wave_format(make<WAVEFORMATEXTENSIBLE>())
+{
+    task_event = CreateEvent(NULL, FALSE, FALSE, NULL);
+    VERIFY(task_event);
+}
+
+PlaybackStreamWasapi::AudioState::~AudioState()
+{
+    CloseHandle(buffer_event);
+}
+
+// The caller has the responsibility to initialize com
+ALWAYS_INLINE AK::Duration PlaybackStreamWasapi::total_time_played_impl(PlaybackStreamWasapi::AudioState* state)
```

Comment:
We could integrate the comment into the function name: `total_time_played_with_com_initialized`